BossMod/QuestBattle/QuestBattleDirector.cs

```diff
@@ -0,0 +1,375 @@
+        var current = true;
+        foreach (var wp in CurrentWaypoints)
+        {
+            Camera.Instance?.DrawWorldLine(start, wp.Position, current ? ArenaColor.Safe : ArenaColor.Danger);
+            current = false;
+            start = wp.Position;
+        }
+    }
+
+    public static bool HaveTarget(Actor player, AIHints hints) => player.InCombat || hints.PriorityTargets.Any(x => hints.Bounds.Contains(x.Actor.Position - hints.Center));
+
+    enum ActionsProhibitedStatus : uint
+    {
+        OutOfTheAction = 2109,
+        InEvent = 1268,
+    }
+
+    private void Dash(Actor player, Vector3 direction, AIHints hints)
+    {
+        if (!_config.UseDash || player.Statuses.Any(s => (ActionsProhibitedStatus)s.ID is ActionsProhibitedStatus.OutOfTheAction or ActionsProhibitedStatus.InEvent || RotationModuleManager.IsRoleplayStatus(s)))
+            return;
+
+        var moveDistance = direction.Length();
+        var moveAngle = Angle.FromDirection(new WDir(direction.XZ()));
+
+        ActionID dashAction = default;
+        var dashDistance = float.MaxValue;
+
+        switch (player.Class)
+        {
+            case Class.PCT:
+                dashAction = ActionID.MakeSpell(PCT.AID.Smudge);
+                dashDistance = 15;
+                break;
+            case Class.DRG:
+                dashAction = ActionID.MakeSpell(DRG.AID.ElusiveJump);
+                dashDistance = 15;
+                moveAngle = Angle.FromDirection(-moveAngle.ToDirection());
+                break;
+            case Class.WHM:
+                dashAction = ActionID.MakeSpell(WHM.AID.AetherialShift);
+                dashDistance = 15;
+                break;
+            case Class.RPR:
+                dashAction = ActionID.MakeSpell(RPR.AID.HellsIngress);
+                dashDistance = 15;
+                break;
+            case Class.DNC:
+                dashAction = ActionID.MakeSpell(DNC.AID.EnAvant);
+                dashDistance = 10;
+                break;
+            case Class.NIN:
+                if (moveDistance > 20)
+                {
+                    var destination = direction / (moveDistance / 20);
+                    hints.ActionsToExecute.Push(ActionID.MakeSpell(NIN.AID.Shukuchi), null, ActionQueue.Priority.Low, targetPos: player.PosRot.XYZ() + destination);
+                }
+                return;
+        }
+
+        if (moveDistance > dashDistance)
+            hints.ActionsToExecute.Push(dashAction, null, ActionQueue.Priority.Low, facingAngle: moveAngle);
+    }
+
+    private async void TryPathfind(Vector3 start, List<Waypoint> connections, int maxRetries = 5)
+    {
+        CurrentConnections = connections;
+        CurrentWaypoints = await TryPathfind(Enumerable.Repeat(new Waypoint(start, false), 1).Concat(connections), maxRetries).ConfigureAwait(false);
+    }
+
+    private async Task<List<NavigationWaypoint>> TryPathfind(IEnumerable<Waypoint> connectionPoints, int maxRetries = 5)
```

Comment:
Can you please explain the async use here? On what thread is it executed, how do you guarantee thread safety?

Reply:
no clue

Reply:
removed async